### PR TITLE
invalid config type provided

### PR DESCRIPTION
**Pull request by wind-c**
examples/persistence/bolt/main.go: invalid config type provided

### CODE REVIEW

File: examples/persistence/bolt/main.go [hunk 1]
```diff
@@ -30,12 +30,15 @@ func main() {
...
 	_ = server.AddHook(new(auth.AllowHook), nil)
 
-	err := server.AddHook(new(bolt.Hook), bolt.Options{
+	err := server.AddHook(new(bolt.Hook), &bolt.Options{
 		Path: "bolt.db",
 		Options: &bbolt.Options{
 			Timeout: 500 * time.Millisecond,
 		},
 	})
+	if err != nil {
```
Comment (mochi-co) [inline examples/persistence/bolt/main.go:39]: Good find

File: server.go [hunk 2]
```diff
@@ -467,7 +467,16 @@ func (s *Server) inheritClientSession(pk packets.Packet, cl *Client) bool {
 			return false // [MQTT-3.2.2-3]
 		}
 
-		cl.State.Inflight = existing.State.Inflight // [MQTT-3.1.2-5]
+		infLen := existing.State.Inflight.Len()
```
Comment (mochi-co) [inline server.go:470]: This is contrary to the mqtt v5 spec:
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Flow_Control

> 4.9 Flow Control [...] The send quota and Receive Maximum value are not preserved across Network Connections, and are re-initialized with each new Network Connection as described above. They are not part of the session state.

and 

> 3.1.2.11.3 Receive Maximum [...] The value of Receive Maximum applies only to the current Network Connection.

**Review by mochi-co — CHANGES_REQUESTED** [review]
Thanks @wind-c ! 
We need the fix for the persistence example for sure - great job seeing it. However, we must not transfer send/receive quota state across connections as it goes against the v5 specification (see comments) and can cause a permanent client deadlock if no acks are returned.

If you can remove the inflight changes to server.go, I'll merge 👍🏻

**Comment by wind-c** [comment]
You can test that when inflight messages are loaded from persistence without resetting quotas, an ErrReceiveMaximum  occurs when client publish message.

File: server.go (hunk 2 above)
Comment (wind-c) [reply server.go:470]: `cl.State.Inflight = existing.State.Inflight`

This code overwrites the quota values of the client to zero, causing an ErrReceiveMaximum when the client publishes the message.

`  
func (s *Server) processPublish(cl *Client, pk packets.Packet) error {
	if !cl.Net.Inline && !IsValidFilter(pk.TopicName, true) {
		return nil
	}

	if atomic.LoadInt32(&cl.State.Inflight.receiveQuota) == 0 {
		return s.DisconnectClient(cl, packets.ErrReceiveMaximum) // ~[MQTT-3.3.4-7] ~[MQTT-3.3.4-8]
	}
`

File: server.go [hunk 3]
```diff
@@ -467,7 +467,16 @@ func (s *Server) inheritClientSession(pk packets.Packet, cl *Client) bool {
...
+		if infLen > 0 {
+			cl.State.Inflight = existing.State.Inflight // [MQTT-3.1.2-5]
+			if cl.State.Inflight.maximumReceiveQuota == 0 && cl.ops.capabilities.ReceiveMaximum != 0 {
+				cl.State.Inflight.ResetReceiveQuota(int32(cl.ops.capabilities.ReceiveMaximum)) // server receive max per client
+				cl.State.Inflight.ResetSendQuota(int32(cl.Properties.Props.ReceiveMaximum))    // client receive max
+				cl.State.Inflight.receiveQuota -= int32(infLen)                                // available receive quota is equal to the maximum quota minus the number of inflight messages
```
Comment (mochi-co) [inline server.go:476]: Ah, sorry for the delay @wind-c - it's been a very busy week. I understand what you are trying to do now you have explained it, and I believe it is correct. The code looks good.

However, my one concern is still this line:

```go
cl.State.Inflight.receiveQuota -= int32(infLen)  // available receive quota is equal to the maximum quota minus the number of inflight messages`
```

Even though there are inflights still pending from the existing, the receive quota should be reset across connections (as specified in the spec). Therefore, we should remove this final line.

What do you think?

Comment (wind-c) [reply server.go:476]: ok, so remove this line.

Comment (mochi-co) [reply server.go:476]: @wind-c Yes, if this line is removed, I believe we can merge this PR. Apologies again for the delay in replying 🙏🏻

**Review by mochi-co — APPROVED** [review]
Looks good to me! Thank you for another excellent contribution, @wind-c

**Pull request merged**